pkg/reconciler/v1alpha1/taskrun/validate_resources.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either extress or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package taskrun
+
+import (
+	"fmt"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/knative/build-pipeline/pkg/reconciler/v1alpha1/taskrun/list"
+	"github.com/knative/build-pipeline/pkg/reconciler/v1alpha1/taskrun/resources"
+)
+
+func validateInputResources(inputs *v1alpha1.Inputs, providedResources map[string]*v1alpha1.PipelineResource) error {
+	if inputs != nil {
+		return validateResources(inputs.Resources, providedResources)
+	}
+	return validateResources([]v1alpha1.TaskResource{}, providedResources)
+}
+
+func validateOutputResources(outputs *v1alpha1.Outputs, providedResources map[string]*v1alpha1.PipelineResource) error {
+	if outputs != nil {
+		return validateResources(outputs.Resources, providedResources)
+	}
+	return validateResources([]v1alpha1.TaskResource{}, providedResources)
+}
+
+func validateResources(neededResources []v1alpha1.TaskResource, providedResources map[string]*v1alpha1.PipelineResource) error {
+	needed := make([]string, 0, len(neededResources))
+	for _, resource := range neededResources {
+		needed = append(needed, resource.Name)
+	}
+	provided := make([]string, 0, len(providedResources))
+	for resource := range providedResources {
+		provided = append(provided, resource)
+	}
+	missing := list.DiffLeft(needed, provided)
+	if len(missing) > 0 {
+		return fmt.Errorf("missing resources: %s", missing)
+	}
+	extra := list.DiffLeft(provided, needed)
+	if len(extra) > 0 {
+		return fmt.Errorf("didn't need these resources by they were provided anyway: %s", extra)
+	}
+	for _, resource := range neededResources {
+		r := providedResources[resource.Name]
+		if resource.Type != r.Spec.Type {
```

Comment:
`r` is a pointer for Resource. Is it better to check for nil pointer? 

```
if r != nil && resource.Type != r.Spec.Type {
```

Reply:
oh boy, thanks for catching that!! I've missed a few of those in the past 😅

Reply:
i guess we should return an error if the resource is missing 🤔 this is a bit odd b/c i think we already have validation that checks fro this earlier - but might as well check it again i guess!

Reply:
oh actually it's a few lines up, these lines should ensure that we never hit that case:

```
	needed := make([]string, 0, len(neededResources))
	for _, resource := range neededResources {
		needed = append(needed, resource.Name)
	}
	provided := make([]string, 0, len(providedResources))
	for resource := range providedResources {
		provided = append(provided, resource)
	}
	missing := list.DiffLeft(needed, provided)
	if len(missing) > 0 {
		return fmt.Errorf("missing resources: %s", missing)
	}
```

I'll add a check for it anyway tho to hopefully make future refactoring easier